sphinx/domains/python.py

```diff
@@ -985,6 +1018,7 @@ class PythonDomain(Domain):
         'classmethod':     PyClassMethod,
         'staticmethod':    PyStaticMethod,
         'attribute':       PyAttribute,
+        'property':        PyProperty,
```

Comment:
Please don’t, properties and attributes should be linked with the same role, as the fact if something is a property or attribute is an implementation detail.

Reply:
Now we are going to add a new objtype "property" to describe properties. It means attributes and properties are not same thing. As discussed in many time, I consider they are similar, but different. So I feel it's natural to add a new role to refer "property" objects. The describing directive and referencing role is a pair.

At same time, as you said, attributes and properties are very similar. So `:py:attr:` role is also able to refer "property" objects. I agree it's natural. But it does not prevent to add `:py:prop:` role. User can use either `:py:attr:` or `:py:prop:` role to refer properties. It is not harmful to add new role for `:py:attr:` role users.

I believe this change does not affect for who uses `py:attribute::` directive and `:py:attr:` to describe and refer to properties. Even after autodoc generates `py:property` directive, users does not need to change their document. `:py:attr:` role will be able to refer properties until now, and from now on. I don't know why you object to add a new role. If you're saying attributes and properties are same, we have to discuss again from how we describe properties.

Reply:
they’re not the same thing to implement, which is why a separate directive makes sense.

they’re the same thing to access which is why only a single way to refer to them makes sense.

Reply:
Okay, please implement as you like. I'll stop to maintenance python domain. It's yours. Bye.

Reply:
I don’t understand. Didn’t you just ask me to discuss this? You said:

> I don't know why you object to add a new role. If you're saying attributes and properties are same, we have to discuss again from how we describe properties.

My comment above was the answer to that:

- I think they **are not** the same on the implementer’s side (the one writing `@property\ndef foo(self):`/`.. property:: foo` **or** `self.foo = ...`/`.. attribute:: foo`)
- and I think they **are** the same on the user’s side (the one writing `instance.foo`/``:attr:`foo` ``)

Reply:
I must apologize to you. I had been frustrated that night. Please forget my comment.

>and I think they are the same on the user’s side (the one writing 

Our opinions still don't match. I think they are different things for users.

* Some properties do not support writing (in other words: assigning or setting)
* Some properties do not support deleting
* Some properties have a side effect (ex. network access, database access, caching)

I understand properties behave like attributes transparently. So But I consider they are similar, but different things. That's why I added `py:property` directive (and why I allow being able to refer a property via `:py:attr:` role). If they are the same, I never add it to the python domain.

I believe there is no reason to add only `py:property` directive. It and `:py:prop:` role must be a pair. If you object to the role, you should object to the directive also.

Reply:
All good :smile: we all have those days. As a fellow maintainer I sure know how it can be. I hope things got better for you!

About the topic: What you describe is all behavior that the library developer wanted to attach to attribute getting/setting. Users just see an attribute that they can get and/or set via what’s called “*attribute* access syntax”. E.g. it could well be that a developer just switched a property to prevent setting an attribute.

I disagree that it’s more consistent to have roles for both compared to the rest of sphinx. There’s a lot of things that can be accessed through e.g. `:meth:` that’s not a instance method. Static methods are functions in reality (as they don’t get bound to `self`), they just happen to be called “methods”. Just pretend “properties” were called “property attributes” or “custom attributes” and my thinking will make sense.

Reply:
>I’m not against .. property::, but just because I don’t think it matters if it’s .. property:: or .. attribute::\n:property:. 

Please let me know why the feature describing a property is needed.  I feel no reason to describe "this attribute is a property" if attributes and properties are the same for users. For library developers?

Reply:
TBH, I'm very tired to discuss properties more. How about removing all supports for properties from python domain? That is nothing different from Sphinx-2.2. I can't understand what is correct.

Reply:
The only benefit of `.. property::` right now is that it supports `:abstractmethod:`. In the future we could also add support for different getter and setter types:

```py
@property
def p(self) -> Path:
    return self._p
@p.setter
def p(self, value: PathLike):
    self._p = Path(p)
```

↓

```restructuredtext
.. property:: p
   :type: ~pathlib.Path
   :setter-type: ~typing.PathLike
```

Reply:
OMG... I've perfectly forgotten about `:abstractmethod:`. To describe such options, I also prefer `py:property` directive. But a discussion about `:py:prop:` role is not concluded yet...